### PR TITLE
Add custom HgpsException, which contains a source_location instance.

### LINKED ISSUE #193: std::source_location in exception messages
I think it's time to add `__FILE__`, `__FUNCTION__`, `__LINE__` to all our exception messages.

Also perhaps we can temporarily catch `out_of_range`, append some info on the vars being accessed, and rethrow. I'm getting a little tired of seeing the useless `std::out_of_range, what: map::at` error messages. 

Something like:

```cpp
Error in __FILE__:__FUNCTION__:__LINE__: exception.what(): some extra debug info.
```

### PR DESCRIPTION
**Pull request by jamesturner246**
Fixes #193 

This change adds a custom exception type `hgps::core::HgpsException`, which is a drop-in replacement for the std exceptions used throughout hgps. This exception stores a new C++20 `std::source_location` object, constructed at the location the exception is constructed, which provides more info in the exception message (line, col, file, function) returned by `HgpsException::what`. For now, only the model_parser and risk factor model code is updated to use the new exceptions to prevent further conflicts with #200.

Limitations:

1) since the `source_location` is constructed at point of `HgpsException` construction, `file()`, `line()`, etcetera, will not exactly pinpoint the point inside a `try` block where, for example, a `map::at` call fails; only where the resulting `out_of_range` exception is caught and rethrown as a `HgpsException`.

2) clang-14 (the default on Ubuntu 22, and used in our CI) does not properly implement `std::source_location`, so some nasty macro stuff is required to substitute for `std::experimental::source_location` for clang <= 14 users.

### CODE REVIEW

**Review by dalonsoa — APPROVED** [review]
It looks good to me! I'm always confused about these things: being a compiled language, how is it possible to get the code location information? Is that contained in the binary somehow?

**Comment by jamesturner246** [comment]
Hummm, that is a good question...

The magic location finding stuff happens in `::current`, and I see there's a note in the [ref](https://en.cppreference.com/w/cpp/utility/source_location/current): 

> std::source_location::current typically requires compiler's built-in implementation. 

so maybe it gets it from the compiler's internal representation directly? not sure.

**Comment by alexdewar** [comment]
I think it probably does require some compiler magic. Debug info can be stripped from binaries later, so that wouldn't be super robust.

Actually, you can see what it's doing here: https://godbolt.org/z/hxh7z41a6

You basically end up with a bunch of hard-coded constants saved into the binary representing the line number etc.

**Review by alexdewar — CHANGES_REQUESTED** [review]
I think this is a great usability improvement. Good work!

One small tweak you could make to the `HgpsException` is to forward on any extra arguments to `fmt::format`. Currently we've got things like this in a few places:

```cpp
throw hgps::core::HgpsException{fmt::format("Failed to parse adjustment file: {} or {}. {}",
                                            male_filename, female_filename, ex.what())};
```

And it would be nicer if you could just write:

```cpp
throw hgps::core::HgpsException{"Failed to parse adjustment file: {} or {}. {}", male_filename, female_filename, ex.what()};
```

You'll need to perfect-forward the arguments, using `std::forward`: https://en.cppreference.com/w/cpp/utility/forward

File: src/HealthGPS.Core/exception.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include <string>
+
+// HACK: Clang 14 does not support std::source_location.
+#if defined(__clang__) && __clang_major__ <= 14
+#include <experimental/source_location>
+using std::experimental::source_location;
+#else
+#include <source_location>
+using std::source_location;
+#endif // defined(__clang__) && __clang_major__ <= 14
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS.Core/exception.h:14]: Fine with me!

**Review by israel-vieira — APPROVED** [review]
Looks good!

File: src/HealthGPS.Core/exception.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS.Core/exception.h:14]: Come to think of it, if this is the only hackery needed to make things work on the
default clang version on Ubuntu 22.04, then maybe we don't need to rush to use a newer clang...

[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS.Core/exception.h:14]: That's what I had in mind, yeah.

File: src/HealthGPS.Core/exception.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#include "exception.h"
+
+#include <fmt/format.h>
+#include <iostream>
+
+namespace hgps::core {
+
+HgpsException::HgpsException(const std::string &what_arg, const source_location location)
+    : std::runtime_error{what_arg}, location_{location} {
+    what_arg_ = fmt::format("{}:{}: {}", file_name(), line(), std::runtime_error::what());
```
Comment (alexdewar) [inline src/HealthGPS.Core/exception.cpp:10]: I've only just noticed this, but how come you're passing `what_arg` to the parent constructor and then constructing a separate formatted string to give the extra source location info? Why doesn't the formatted string include `what_arg`? Surely we want this message to be visible to users. Why don't you construct a formatted string containing `what_arg` and the source location info, pass it to `std::runtime_error`'s constructor and drop the custom `what()` method?

**Review by alexdewar — APPROVED** [review]
LGTM!

**Pull request merged**